Add support for overriding default CUD opperations

jasper-lyons opened the pull request:
This follows on from https://github.com/rom-rb/rom-repository/issues/68.

I would like some feedback on how to improve the testcase and bring it more inline with the existing tests if possible!

jasper-lyons committed:
Add support for overriding default CUD opperations

jasper-lyons committed:
Add support for overriding default CUD opperations

solnic commented on lib/rom/repository/command_proxy.rb:
```diff
@@ -11,11 +11,10 @@ class CommandProxy
       def call(input)
-        command.call(root => input)
+        command.call(command.name.relation => input)
       end
```
Hmm I think this change won't cause issues but I'm wondering what made you change this?

jasper-lyons replied:
The reason is because the previous value was:

`Dry::Core::Inflector.singularize(command.name.relation).to_sym`

Which meant, if the command was for relation `:users` the command was being called with:

`command.call(:user => { ... })`

But the `tuple_path` in the input elevator called when executing a number of the commands was looking for `:users`. Changing it to `command.name.relation` instead meant that the commands were always called like:

`command.call(users: { ... })`

Which provided the right key for the `tuple_path` in the `InputEvaluator` to be traversed.

I imagine it would have been better to update how the `tuple_path` was constructed but I have 0 idea how to do that.

solnic replied:
Right, makes sense not to do something that's not really needed :D

This class is a temporary "solution", the only reason why it exists is because command graphs in rom core, as you noticed, need the relation name as the key. If we can tweak it so this is no longer needed, then this whole proxy can be removed.

jasper-lyons replied:
Reading that code comment made me feel a little better about changing this class. I think we're not far off making this obsolete. Could you point me in the direction of some cool bits of the command graph code to read?

solnic replied:
It is probably not that difficult to tweak the code in core so that we can ask graph builder to build the command where the first one is marked as a root without a tuple path. Basically [here](https://github.com/rom-rb/rom/blob/master/lib/rom/commands/graph/class_interface.rb#L44) we ALWAYS append a key to the path (and path is [empty by default](https://github.com/rom-rb/rom/blob/master/lib/rom/commands/graph/class_interface.rb#L21)) that's how we end up with evaluator that always expects a tuple path. So we'd have to change two things:

1) make it possible to do something like `ROM::Commands::Graph.build(registry, graph_opts, false` which we'd call right [here](https://github.com/rom-rb/rom-repository/blob/master/lib/rom/repository/command_compiler.rb#L46)
2) tweak these [two lines](https://github.com/rom-rb/rom/blob/master/lib/rom/commands/graph/class_interface.rb#L44-L45) so that if `path` is `false` then we use an input evaluator that will simply return the input (literally `-> input { input }`)

jasper-lyons replied:
Perhaps if the path is empty, `InputEvaluator` should just return the input? Then we could just not add the key to the path?

I don't understand how the path is built up though.

solnic replied:
@jasper-lyons yeah it crossed my mind to just allow `nil` but it's a bit nasty, because we'd be adding special meaning to `nil` and IF it's `nil` when it shouldn't be we'd be wondering why and debugging would be harder. So I'd say it's better to use input evaluator when there's actually something to evaluate, otherwise just pass the input as it is.

> I don't understand how the path is built up though.

It's built by traversing through the command graph ast and building it up as we dive deeper, see [the spec](https://github.com/rom-rb/rom/blob/master/spec/integration/commands/graph_spec.rb#L94-L119) where `option` is a sample ast, we have `:users` then `:tasks` so we end up with `:'users.tasks'` tuple path while traversing through these options.

jasper-lyons replied:
> So I'd say it's better to use input evaluator when there's actually something to evaluate, otherwise just pass the input as it is.

Completely agree. Lets do that.

> tweak these two lines so that if `path` is `false` then we use an input evaluator that will simply return the input (literally -> input { input })

Why not when path is empty?

solnic replied:
> Why not when path is empty?

Because it would break public API

jasper-lyons replied:
OK yeah, I see that.

Kukunin commented on .gitignore:
```diff
@@ -1,3 +1,5 @@
 log/*.log
 doc
+.byebug_history
+tags
```
Vim or Emacs? 
BTW, you can configure global `gitignore` for all projects ever to ignore such files

jasper-lyons replied:
Vim.

I should definitely do that. Thanks.

jasper-lyons commented on spec/integration/override_create_spec.rb:
```diff
@@ -0,0 +1,38 @@
+      end
+    end
+
+    class CreateAccount < ROM::Commands::Create[:memory]
```
If I could workout how to do this without creating a stand alone constant I could bring this code more inline with the existing tests. @solnic?

solnic replied:
you can use `configuration` to define a command using rom's DSL:

``` ruby
configuration.commands(:users) do
  define(:create)
end
```

jasper-lyons committed:
Clean up overriding CUD command test.

I use a cludgy `Object.send(:remove_const, :CustomCommand)` here
to allow the naming of the command object which make checking
the type of the command returned easy. This could be implemented
with some duck typing perhaps?

jasper-lyons committed:
Clean up overriding CUD command test.

I use a cludgy `Object.send(:remove_const, :CustomCommand)` here
to allow the naming of the command object which make checking
the type of the command returned easy. This could be implemented
with some duck typing perhaps?

jasper-lyons committed:
Remove machine specific .gitignore.

Moved it into my global ignore instead.

jasper-lyons committed:
Remove machine specific .gitignore.

Moved it into my global ignore instead.

jasper-lyons committed:
Fix failing test by including id in query

jasper-lyons committed:
Fix failing test by including id in query

jasper-lyons committed:
Merge branch 'master' of github.com:releaseplatform/rom-repository

jasper-lyons committed:
Merge branch 'master' of github.com:rom-rb/rom-repository
